core/src/socket.rs

```diff
@@ -111,100 +160,238 @@ impl<'gc> Sockets<'gc> {
+                                .classes()
+                                .progressevent
+                                .construct(
+                                    &mut activation,
+                                    &[
+                                        "socketData".into(),
+                                        false.into(),
+                                        false.into(),
+                                        bytes_loaded.into(),
+                                        //NOTE: bytesTotal is not used by socketData event.
+                                        0.into(),
+                                    ],
+                                )
+                                .expect("ProgressEvent should be constructed");
+
+                            Avm2::dispatch_event(
+                                &mut activation.context,
+                                progress_evt,
+                                target.into(),
+                            );
+                        }
+                        SocketKind::Avm1(target) => {
+                            let mut activation = Avm1Activation::from_stub(
+                                context.reborrow(),
+                                ActivationIdentifier::root("[XMLSocket]"),
+                            );
+
+                            // NOTE: This is enforced in connect_avm1() function.
+                            let xml_socket =
+                                XmlSocket::cast(target.into()).expect("target should be XmlSocket");
+
+                            // Get XmlSocket read buffer.
+                            let mut buffer = xml_socket.read_buffer();
+                            buffer.extend(data);
+
+                            let mut messages = vec![];
+
+                            // Check for messages.
+                            while let Some((index, _)) =
+                                buffer.iter().enumerate().find(|(_, &b)| b == 0)
+                            {
+                                let message = buffer.drain(..index).collect::<Vec<_>>();
+                                // Remove null byte.
+                                let _ = buffer.drain(..1);
+
+                                // Store the message in cache.
+                                messages.push(AvmString::new_utf8_bytes(activation.gc(), &message));
+                            }
+
+                            // Drop the reference to the buffer, to make sure the SWF can call with close.
+                            drop(buffer);
+
+                            for message in messages {
+                                let _ = target.call_method(
+                                    "onData".into(),
+                                    &[message.into()],
+                                    &mut activation,
+                                    ExecutionReason::Special,
+                                );
+                            }
```

Comment:
I was think of a better way to process messages without using 2 loops and an extra Vec, what about this code? I think it is more efficient as it only consider null bytes in `data`, in order to have a complexity of `O(data.len())` instead of `O(read_buffer.len())`

```suggestion
                            let mut data = data;

                            if let Some((index, _)) = data.iter().enumerate().find(|(_, &b)| b == 0)
                            {
                                // There's no null byte in the read buffer so we can be sure
                                // all remaining bytes will be consumed by the next message.
                                let mut buffer = xml_socket
                                    .read_buffer()
                                    .drain(..)
                                    .chain(data.drain(..index))
                                    .collect::<Vec<_>>();

                                loop {
                                    // Remove null byte.
                                    let _ = data.drain(..1);

                                    let message =
                                        AvmString::new_utf8_bytes(activation.gc(), &buffer);

                                    let _ = target.call_method(
                                        "onData".into(),
                                        &[message.into()],
                                        &mut activation,
                                        ExecutionReason::Special,
                                    );

                                    if let Some((index, _)) =
                                        data.iter().enumerate().find(|(_, &b)| b == 0)
                                    {
                                        // We already consumed everything in the buffer so arriving here
                                        // means the next message is fully in data only, no need to access
                                        // the buffer again then.
                                        buffer = data.drain(..index).collect::<Vec<_>>();
                                    } else {
                                        // No full message left in data, we'll wait for the next chunk
                                        // from the server.
                                        break;
                                    }
                                }
                            };

                            // Append what remain to XmlSocket read buffer.
                            // Since we consumed all null byes above, we can be sure
                            // the read buffer will never contain one.
                            if !data.is_empty() {
                                xml_socket.read_buffer().extend(data);
                            }

```

Obviously it's just a suggestion, your code works too :+1:

Reply:
It does hinder code readability, but I think the  current way is performant enough.

Reply:
I thought about this, for a bit and implemented your suggestion in [8037eff](https://github.com/ruffle-rs/ruffle/pull/12489/commits/8037eff6e3755ed88dcfa9e164b29765cfbc6046)